apps/emqx_bridge_iceberg/src/emqx_bridge_iceberg_impl.erl

```diff
@@ -0,0 +1,775 @@
+            ?table_uuid => TableUUID,
+            ?write_uuid => WriteUUID
+        },
+        {ok, TransferState, ContainerOpts}
+    end.
+
+-spec process_append(iodata(), write_metadata(), transfer_state()) ->
+    transfer_state().
+process_append(IOData, WriteMetadata, TransferState0) ->
+    #{num_records := N1} = WriteMetadata,
+    #{
+        ?num_records := N0,
+        ?data_size := S0,
+        ?s3_transfer_state := S3TransferState0
+    } = TransferState0,
+    S1 = iolist_size(IOData),
+    {ok, S3TransferState} = emqx_s3_upload:append(IOData, S3TransferState0),
+    TransferState0#{
+        ?num_records := N0 + N1,
+        ?data_size := S0 + S1,
+        ?s3_transfer_state := S3TransferState
+    }.
+
+-spec process_write(transfer_state()) ->
+    {ok, transfer_state()} | {error, term()}.
+process_write(TransferState0) ->
+    #{
+        ?s3_transfer_state := S3TransferState0
+    } = TransferState0,
+    case emqx_s3_upload:write(S3TransferState0) of
+        {ok, S3TransferState} ->
+            TransferState = TransferState0#{
+                ?s3_transfer_state := S3TransferState
+            },
+            {ok, TransferState};
+        {cont, S3TransferState} ->
+            TransferState = TransferState0#{
+                ?s3_transfer_state := S3TransferState
+            },
+            process_write(TransferState);
+        {error, Reason} ->
+            _ = emqx_s3_upload:abort(S3TransferState0),
+            {error, Reason}
+    end.
+
+-spec process_complete(transfer_state()) ->
+    {ok, term()} | no_return().
+process_complete(TransferState0) ->
+    #{
+        ?s3_transfer_state := S3TransferState
+    } = TransferState0,
+    case emqx_s3_upload:complete(S3TransferState) of
+        {ok, _S3Completed} ->
+            ?tp("iceberg_upload_manifests_enter", #{}),
+            upload_manifests(TransferState0);
+        {error, Reason} ->
+            _ = emqx_s3_upload:abort(S3TransferState),
+            exit({upload_failed, {data_file, Reason}})
+    end.
+
+%%------------------------------------------------------------------------------
+%% Internal fns
+%%------------------------------------------------------------------------------
+
+make_client(#{parameters := #{location_type := s3tables} = Params}) ->
+    emqx_bridge_iceberg_client_s3t:new(Params).
+
+init_location_client(ConnResId, #{parameters := #{location_type := s3tables} = Params}) ->
+    maybe
+        #{s3_client := S3Config0} = Params,
+        S3Config = S3Config0#{url_expire_time => 0},
+        S3ClientConfig = emqx_s3_profile_conf:client_config(S3Config, ConnResId),
+        _ = emqx_s3_client_http:stop_pool(ConnResId),
+        ok ?= emqx_s3_client_http:start_pool(ConnResId, S3Config),
+        {ok, #{?s3_client_config => S3ClientConfig}}
+    end.
+
+run_aggregated_action(Batch, #{?aggreg_id := AggregId}) ->
+    Timestamp = erlang:system_time(second),
+    case emqx_connector_aggregator:push_records(AggregId, Timestamp, Batch) of
```

Comment:
TODO: should we validate the records with the current known schema before pushing?
Schema can change later, anyways.  :thinking: